collector/vrrp.go

```diff
@@ -46,83 +59,26 @@ type VrrpInstanceStats struct {
 }
 
-// VRRPCollector collects VRRP metrics, implemented as per prometheus.Collector interface.
-type VRRPCollector struct{}
-
-// NewVRRPCollector returns a VRRPCollector struct.
-func NewVRRPCollector() *VRRPCollector {
-	return &VRRPCollector{}
-}
-
-// Name of the collector. Used to populate flag name.
-func (*VRRPCollector) Name() string {
-	return vrrpSubsystem
-}
-
-// Help describes the metrics this collector scrapes. Used to populate flag help.
-func (*VRRPCollector) Help() string {
-	return "Collect VRRP Metrics"
-}
-
-// EnabledByDefault describes whether this collector is enabled by default. Used to populate flag default.
-func (*VRRPCollector) EnabledByDefault() bool {
-	return false
-}
-
-// Describe implemented as per the prometheus.Collector interface.
-func (*VRRPCollector) Describe(ch chan<- *prometheus.Desc) {
-	for _, desc := range getVRRPDesc() {
-		ch <- desc
-	}
-}
-
-// Collect implemented as per the prometheus.Collector interface.
-func (c *VRRPCollector) Collect(ch chan<- prometheus.Metric) {
-	collectVRRP(ch)
-}
-
-// CollectErrors returns what errors have been gathered.
-func (*VRRPCollector) CollectErrors() []error {
-	return vrrpErrors
+type vrrpCollector struct {
+	logger log.Logger
 }
 
-// CollectTotalErrors returns total errors.
-func (*VRRPCollector) CollectTotalErrors() float64 {
-	return totalVRRPErrors
+// NewVRRPCollector collects VRRP metrics, implemented as per the Collector interface.
+func NewVRRPCollector(logger log.Logger) (Collector, error) {
+	return &vrrpCollector{logger: logger}, nil
 }
 
-func getVRRPDesc() map[string]*prometheus.Desc {
-	if vrrpDesc != nil {
-		return vrrpDesc
-	}
-
-	vrrpLabels := []string{"proto", "vrid", "interface", "subinterface"}
-	vrrpStateLabels := append(vrrpLabels, "state")
-
-	vrrpDesc = map[string]*prometheus.Desc{
-		"vrrpState":       colPromDesc(vrrpSubsystem, "state", "Status of the VRRP state machine.", vrrpStateLabels),
-		"adverTx":         colPromDesc(vrrpSubsystem, "adverTx_total", "Advertisements sent total.", vrrpLabels),
-		"adverRx":         colPromDesc(vrrpSubsystem, "adverRx_total", "Advertisements received total.", vrrpLabels),
-		"garpTx":          colPromDesc(vrrpSubsystem, "garpTx_total", "Gratuitous ARP sent total.", vrrpLabels),
-		"neighborAdverTx": colPromDesc(vrrpSubsystem, "neighborAdverTx_total", "Neighbor Advertisements sent total.", vrrpLabels),
-		"transitions":     colPromDesc(vrrpSubsystem, "state_transitions_total", "Number of transitions of the VRRP state machine in total.", vrrpLabels),
-	}
-
-	return vrrpDesc
-}
-
-func collectVRRP(ch chan<- prometheus.Metric) {
-	vrrpErrors = []error{}
+// Update implemented as per the Collector interface.
+func (c *vrrpCollector) Update(ch chan<- prometheus.Metric) error {
 	jsonVRRPInfo, err := getVRRPInfo()
 	if err != nil {
-		totalVRRPErrors++
-		vrrpErrors = append(vrrpErrors, fmt.Errorf("cannot get vrrp info: %s", err))
+		return fmt.Errorf("cannot get vrrp info: %s", err)
```

Comment:
Consider wrapping errors using the `%w` verb where you include it in a `fmt.Errorf()`, e.g.
```suggestion
		return fmt.Errorf("cannot get vrrp info: %w", err)
```

Reply:
Fixed.